### ISSUE TITLE
Native Docker Interface

### ISSUE THREAD

**Issue by will-chan** [rancherio/rancher#47]
Although Rancher does provide a set of API to manipulate Docker containers, it also needs the ability to support the following:

* Poll mechanism for Rancher agents to detect existing containers running on a host and import them into Rancher.
* Event listening mechanism for Rancher agents to detect new changes from Docker API execution.  (i.e. a user executing a docker run command locally on a host, adding new ports, etc.).

**Comment by cjellick** [comment]
Notes from disscussion with @ibuildthecloud:

### Stories/use cases/functionality
#### Allow containers created outside of Rancher to be orchestrated by Rancher.
- [ ] As a user, I should be able to (ISBAT) start a container outside of Rancher but still have Rancher manage my container's networking.
- [ ] As a user, ISBAT to start a container outside of Rancher and still have that container participate in Rancher service discovery and load balancing.
- [ ] As a user, ISBAT to perform all actions on a container that was created outside of Rancher. Examples:
  - stop, restart, start
  - view logs, container exec, connect to console, view stats
- [ ] As a user, I want the name I give a container in Rancher to be reflected in Docker (Rancher currently names the container as a UUID in Docker)

### Related bugs that should be fixed as part of this effort
- [ ] net-util appears to error out if a container is extremely short lived
- 

### General backend changes needed
- [ ] Implement

**Comment by cjellick** [comment]
Here's a list of related existing issues that should be addressed as part of this effort:
- [ ] https://github.com/rancherio/rancher/issues/13 - net-util related
- [ ] https://github.com/rancherio/rancher/issues/23 - net-util related
- [ ] https://github.com/rancherio/rancher/issues/35 - net-util related
- [ ] https://github.com/rancherio/cattle/issues/199 - Naming containers
- [ ] https://github.com/rancherio/cattle/issues/136 - Flagging rancher management containers


And these issues are not directly related, but I want to reference just in case I end up addressing them as a side effect:
- [ ] https://github.com/rancherio/cattle/issues/230 - Host state
- [ ] https://github.com/rancherio/python-agent/issues/37 - Supporting docker network modes
- [ ] https://github.com/rancherio/python-agent/issues/30 - Support network disabled
- [ ] https://github.com/rancherio/python-agent/issues/14 - Volume management

**Comment by cjellick** [comment]
How will we handle weird circular scenarios like this:

A user creates container outside of rancher with `--net=rancher:..` and some links specified like: `--link=foo:bar`.

We currently have special logic to remove the link as a docker create parameter and instead emulate links by injecting environment variables. If we can't intercept the create command, then our currently link solution wont work.

**Comment by cjellick** [comment]
Talked with @ibuildthecloud and a few things came up:
1. Defer native docker run + cross server linking until later. Not the most important
2. There's an additional "general backend change" item that's need:
 - Fully implement and refine the "ping" logic that will import containers if rancher is brought up on a pre-existing host. 
3. Our aggressive target is that in 3 weeks, *3/18* we have a basic docker run use case deployed. The functionality may be incomplete and heavily caveated, but a user should be able to do a `docker run...` from a host and have that container show up in Rancher. One of the big caveats is that this probably will only support unmanaged networking.

Ping @will-chan.

**Comment by cjellick** [comment]
After discussing some swarm-integration design decisions with @will-chan @sonchang and @ibuildthecloud, we came away with the following:

Swarm needs Native Docker Interface (NDI) to handle the wiring up of managed networking. This means full swarm integration is blocked on NDI.

Full-on NDI won't be done for several weeks, so this is a big blocker for swarm. Darren and I discussed this and realized that if we change our approach to fixing the net-util idempotency issues, we can provide managed networking support for swarm and fix net-utils at the same time and in the nearer term (targeting sometime next week, the week of **3/4**).

So here's the new approach, what I'll start working on immediately:
* *Background:* the net-utils logic that currently lives in python-agent as part of the instace_activate flow uses nsenter to get into the container's networking namespace and configure the rancher IP address on the appropriate device in the container. So, it needs the container to be local to work properly.
* We're going to remove net-utils logic completely from python-agent and move it to host-api. We'll write a small docker-event-listener in go in host-api and do the ip injection from there.
* To accomplish this, we'll add a "label" (for now it'll be implemented as an evironment variable because docker doesnt support labels) to the container that is the rancher ip address. When the docker-event-listener sees a create or start event, it will inspect the container for that label. If it finds it, it will execute the net-utils logic.

Further implementation details, mainly for my own reference:
* The net-util logic must be idempotent so that it can rerun on create and start events without breaking.
* Keep the event logic self contained inside a package in host-api so that it can be broken out into a separate service at a later date.
* Need logic that inspects **all** containers for the label when host-api starts up, in case it went down.

**Comment by cjellick** [comment]
Pull requests to address the net-util bugs and refactor:
https://github.com/rancherio/cattle/pull/299
https://github.com/rancherio/python-agent/pull/90
https://github.com/rancherio/validation-tests/pull/26
https://github.com/rancherio/host-api/pull/17

**Comment by cjellick** [comment]
@ibuildthecloud The net-util refactor is ready for review. It is encompassed by the above 4 pull requests.
Merging these will require some coordination, so I'll write it up here.

https://github.com/rancherio/validation-tests/pull/26 can be merged immediately. It just checks that IPs are getting configured correctly inside the VMs. It is meant to prove that behavior doesn't change/break between the old-way and new-way.

https://github.com/rancherio/host-api/pull/17 can be merged at will. It is backwards compatible since it will not do anything if the container doesn't have the RANCHER_IP env var.

https://github.com/rancherio/python-agent/pull/90 will cause python-agent to stop running net-util and instead add the RANCHER_IP env varn to the container. The host-api change should be merged, released, and it's version updated in cattle before this one is merged.

https://github.com/rancherio/cattle/pull/299 should fail until everything is merged and released. Its goal is to prove that the old way of doing net-util was broken and that the new way will fix it.

**Comment by cjellick** [comment]
For importing an existing host, I dont plan to automatically give them rancher networking. Would not be easy to determine if that was the desire behavior.

**Comment by eagoweb** [comment]
Hi @cjellick is it working in the current master? In that case how can I test it?

**Comment by vincent99** [comment]
@eagoweb Yes this is in the last couple releases.. Open up a console to one of your hosts and the UI in a browser.  The simplest test is doing `docker run <something>` and the container will automatically show up in Rancher, but stop/start/etc also work.  You can also add a host that has existing containers to Rancher and they will be "imported".

**Comment by Niceplace** [comment]
@vincent99 @cjellick This doesn't seem to work for me, I'm running the rancher agent on the same host as the rancher server.

What I see on the server : 

![image](https://cloud.githubusercontent.com/assets/8710189/23183167/1dbb05c0-f849-11e6-8219-1c761252711d.png)

What I see on my host (docker ps -a)

![image](https://cloud.githubusercontent.com/assets/8710189/23183222/50a5dde8-f849-11e6-9a17-ba6a2a9084c2.png)

Status of the host seems OK :

![image](https://cloud.githubusercontent.com/assets/8710189/23183278/82a92eda-f849-11e6-9a3e-32f7e1abd4dc.png)

**Comment by cjellick** [comment]
@Niceplace would you mind opening a new issue for this? 

Ping me in the issue.
Also, here's a few follow up questions maybe you can answer in the new issue:
- if you deploy a container from the UI, does it come up successfully?
- if you do a simple `docker run -d ubuntu` from the host, does it show up in the UI? If not, does refeshing the UI cause it to show up?
